### PR TITLE
Weighted Cross Entropy Implementation

### LINKED ISSUE #141: Extend loss weight calculation for classification
**Is your feature request related to a problem? Please describe.**
Currently, classification problems don't use weight calculations for specific classes. 

**Describe the solution you'd like**
Similar to segmentation, the weights should be dynamic.

**Describe alternatives you've considered**
N.A.

**Additional context**
N.A.

### PR DESCRIPTION
**Pull request by orhunguley**
Fixes #141

## Proposed Changes

- Added weighted cross entropy loss to losses.py

## Checklist

- [x] I have read the [`CONTRIBUTING`](https://github.com/CBICA/GaNDLF/blob/master/CONTRIBUTING.md) guide
- [x] My PR is based from the [current GaNDLF master ](https://garygregory.wordpress.com/2016/11/10/how-to-catch-up-my-git-fork-to-master/)
- [x] Non-breaking change (would not break existing functionality): if changes breaks current code, please provide as many details as possible
- [x] Function/class source code documentation added/updated
- [x] Code has been [blacked](https://github.com/psf/black#usage) for style consistency
- [ ] If applicable, version information [has been updated in GANDLF/version.py](https://github.com/CBICA/GaNDLF/blob/master/GANDLF/version.py)
- [ ] If adding a submodule, add to list of expections for black styling in [pyproject.toml](https://github.com/CBICA/GaNDLF/blob/master/pyproject.toml) file 
- [ ] [Usage documentation](https://github.com/CBICA/GaNDLF/blob/master/docs) has been updated, if appropriate
- [ ] [History](https://github.com/CBICA/GaNDLF/blob/master/HISTORY.md) has been updated, if appropriate
- [ ] Tests added or modified to [cover the changes](https://app.codecov.io/gh/CBICA/GaNDLF); if coverage is reduced, please give explanation

### CODE REVIEW

**Review by sarthakpati — CHANGES_REQUESTED** [review]
Hey @orhunguley,

This is a nice PR!

I have 2 suggestions:
1. https://github.com/CBICA/GaNDLF/pull/151#discussion_r676074077
2. The functions `cel` and `weighted_cel` don't need to be separate. Put a single function and pass `params["class_weights"]` (`params["class_weights"]=None` if weights are not calculated [[ref](https://github.com/orhunguley/GaNDLF/blob/master/GANDLF/training_loop.py#L729)], so it should be fine) onward to `torch.nn.CrossEntropyLoss()`. This will reduce the code complexity (and also ensure that the function is tested by the unit tests.

Cheers!

File: GANDLF/utils.py [hunk 1]
```diff
@@ -506,60 +497,70 @@ def get_class_imbalance_weights(training_data_loader, parameters):
...
+            value_to_predict = subject["value_0"][0]
+            for i in range(0, len(parameters["model"]["class_list"])):
+                if value_to_predict == i:
+                    abs_dict[i] += 1
+                    # we only want to increase the counter for those subjects that are defined in the class_list
+                    total_counter += 1
+
+    # Normalize class weights
+    normalizer = sum(abs_dict.values())
+    weights_dict = {key: val / normalizer for key, val in abs_dict.items()}
```
Comment (sarthakpati) [inline GANDLF/utils.py:547]:
```suggestion
    weights_dict = {key: val / total_counter for key, val in abs_dict.items()}
```

File: GANDLF/losses.py [hunk 2]
```diff
@@ -14,12 +14,23 @@ def dice(inp, target):
...
+    if params["class_weights"]:
+        class_weights = torch.FloatTensor(list(params["class_weights"].values()))
+
+        # more examples you have in the training data, the smaller the weight you have in the loss
+        class_weights = 1.0 / class_weights
+
+        if out.is_cuda:
+            class_weights = class_weights.cuda()
+        else:
+            class_weights = class_weights.cpu()
```
Comment (sarthakpati) [inline GANDLF/losses.py:30]:
```suggestion
        class_weights .float().to(params["device"])
```

**Review by sarthakpati — APPROVED** [review]
Needed a few more fixes, which I have pushed.

**Comment by orhunguley** [comment]
Hi @sarthakpati  
I am not sure if removing cuda() check for "out" in weighted cel calculation would work. Because even though we have the device as "cuda", the validation is all cpu based and would cause errors. I will check it and let you know.

**Comment by sarthakpati** [comment]
Sure thing. I checked on my location machine for both CPU and GPU and it did seem to work as expected.

**Comment by orhunguley** [comment]
> Sure thing. I checked on my local machine for both CPU and GPU and it did seem to work as expected.

I got an error. Can you also try it on your GPU machine? Maybe you did not wait until validation, that is why you did not encounter any errors?

**Comment by sarthakpati** [comment]
Understood. Put in a simple fix: #153

Thanks!
